source/common/router/rds_impl.cc

```diff
@@ -211,6 +222,37 @@ void RdsRouteConfigProviderImpl::onConfigUpdate() {
     prev_config->config_ = new_config;
     return previous;
   });
+
+  const auto aliases = config_update_info_->aliasesInLastVhdsUpdate();
+  // Regular (non-VHDS) RDS updates don't populate aliases fields in resources.
+  if (aliases.empty()) {
+    return;
+  }
+
+  // Notifies connections that RouteConfiguration update has been propagated.
+  // Callbacks processing is performed in FIFO order and stops when the alias(es) used in the VHDS
+  // update request do not match the aliases in the update response. The assumption is the response
+  // to a request is not going to arrive before all requests before it have been responded to.
+  config_update_callbacks_->runOnAllThreads(
+      [aliases](ThreadLocal::ThreadLocalObjectSharedPtr previous)
+          -> ThreadLocal::ThreadLocalObjectSharedPtr {
+        auto callbacks = std::dynamic_pointer_cast<ThreadLocalCallbacks>(previous)->callbacks_;
+        std::vector<std::string> aliases_not_in_update;
+        while (!callbacks.empty()) {
+          auto update_on_demand_callback = callbacks.front();
+          std::set_difference(update_on_demand_callback.aliases_.begin(),
+                              update_on_demand_callback.aliases_.end(), aliases.begin(),
+                              aliases.end(), std::back_inserter(aliases_not_in_update));
+          if (aliases_not_in_update.empty()) {
+            callbacks.pop();
+            update_on_demand_callback.cb_();
+            aliases_not_in_update.clear();
+          } else {
+            break;
```

Comment:
There are many Http requests who could fire a ondemand request, think about this sequence of pending requests:

request1: {hostA}
request2: {hostB}
request3...N: {hostA}

Now RDS updated comes back with {host A}, this while loop will ignore requests from request {3...N}, they'd hang waiting for updates?
I think you need to go-over the whole callback list.

Reply:
done